### PR TITLE
561 create mutation for duplicating a page

### PR DESCRIPTION
**Pull request by K9Morphed**
### What is the context of this PR?
This PR adds the functionality in order to duplicate a page. 

>It also includes code for duplicating a section and questionnaire; these functions will not work just yet because some implementation for duplicating a page breaks these. This should be fixed in separate PRs because the additional functionality is not in scope for this ticket.

### How to review 
All tests should pass, and the code should work when calling the mutations included in [this schema PR](https://github.com/ONSdigital/eq-author-graphql-schema/pull/57)

### CODE REVIEW

File: repositories/PageRepository.test.js [hunk 1]
```diff
@@ -440,4 +446,73 @@ describe("PagesRepository", () => {
...
+      const pageOne = await PageRepository.insert(
+        buildPage({ sectionId: section.id })
+      );
+
+      const pageTwo = await PageRepository.copyPage(pageOne.id, 1);
+
+      const positionOfPageOne = await PageRepository.getPosition(pageOne);
+      const positionOfPageTwo = await PageRepository.getPosition(pageTwo);
+
+      expect(parseInt(positionOfPageOne) + 1).toEqual(
```
Comment (adammahmood) [inline repositories/PageRepository.test.js]: It's a bit annoying that we have to use `parseInt` on the `position` values all the time.  I have created a Tech Debt ticket to investigate this: https://trello.com/c/r1NbxAwr/312-review-bigint-usage-in-api

File: repositories/strategies/duplicateStrategy.js [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  if (!isNil(parentRelation)) {
+    duplicatedRecord[get(parentRelation, "columnName")] = get(
+      parentRelation,
+      "id"
+    );
+  }
+
+  const newRecord = { ...duplicatedRecord, ...other };
+
+  return Promise.resolve(
```
Comment (adammahmood) [inline repositories/strategies/duplicateStrategy.js]: The `Promise.resolve` is unnecessary.  This pattern also appears frequently in this file.

File: repositories/strategies/duplicateStrategy.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
+const { omit, head, get, isNil } = require("lodash");
```
Comment (adammahmood) [inline repositories/strategies/duplicateStrategy.js]: This file should be unit tested.

File: repositories/strategies/duplicateStrategy.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+  return returnedData;
+};
+
+const selectData = async (trx, tableName, columns, where, callback) => {
+  return trx
+    .select(columns)
+    .from(tableName)
+    .where(where)
+    .then(callback);
```
Comment (adammahmood) [inline repositories/strategies/duplicateStrategy.js]: Unnecessary callback.  The promise is returned so the caller can add their own `.then()`

File: schema/resolvers.js [hunk 5]
```diff
@@ -47,6 +47,8 @@ const Resolvers = {
       ctx.repositories.Questionnaire.remove(args.input.id),
     undeleteQuestionnaire: (_, args, ctx) =>
       ctx.repositories.Questionnaire.undelete(args.input.id),
+    duplicateQuestionnaire: (_, args, ctx) =>
```
Comment (adammahmood) [inline schema/resolvers.js]: We should add a resolver test for the new functions in this file.

File: repositories/PageRepository.test.js (hunk 1 above)
Comment (K9Morphed) [reply repositories/PageRepository.test.js]: @adammahmood Alright. I'll resolve this conversation for now and then we can revisit this once the TechDebt card has been completed, if you agree? I don't feel that it should be a blocker for this task.

File: repositories/PageRepository.js [hunk 6]
```diff
@@ -1,7 +1,9 @@
 const { map, head, invert, get } = require("lodash/fp");
+const replaceTitle = require("../utils/updateTitle");
```
Comment (samiwel) [inline repositories/PageRepository.js]: Name of function and file should ideally be the same.

File: repositories/PageRepository.js [hunk 7]
```diff
@@ -95,6 +97,20 @@ function getRoutingDestinations(pageId) {
   return db.transaction(trx => getAvailableRoutingDestinations(trx, pageId));
 }
 
+function copyPage(id, position) {
```
Comment (samiwel) [inline repositories/PageRepository.js]: Rename to `duplicatePage` so that our nomenclature is consistent?

File: repositories/PageRepository.js [hunk 8]
```diff
@@ -95,6 +97,20 @@ function getRoutingDestinations(pageId) {
...
+  return db.transaction(async trx => {
+    const page = await trx
+      .select("*")
+      .from("Pages")
+      .where({ id: id })
```
Comment (samiwel) [inline repositories/PageRepository.js]:
```js
.where({ id })
```

File: repositories/QuestionnaireRepository.js [hunk 9]
```diff
@@ -3,6 +3,24 @@ const Questionnaire = require("../db/Questionnaire");
 const mapFields = require("../utils/mapFields");
 const mapping = { created_at: "createdAt" }; // eslint-disable-line camelcase
 const fromDb = mapFields(mapping);
+const db = require("../db");
+const {
+  duplicateQuestionnaireStrategy
+} = require("./strategies/duplicateStrategy");
+
+module.exports.copyQuestionnaire = function(id) {
```
Comment (samiwel) [inline repositories/QuestionnaireRepository.js]: Same as above inconsistent naming

File: utils/updateTitle.js [hunk 10]
```diff
@@ -0,0 +1,6 @@
+const { replace, isNil } = require("lodash");
+
+module.exports = (title, prefix = "Copy of") =>
+  replace(title, /(<\w*>)?(.*)/, (_, a, b) => {
+    return isNil(a) ? `${prefix} ${b}` : `${a}${prefix} ${b}`;
+  });
```
Comment (samiwel) [inline utils/updateTitle.js]: This would benefit from some unit tests. Be good to test passing a title surrounded by tags, containing tags and without tags so behaviour is documented by example.

File: repositories/PageRepository.test.js (hunk 1 above)
Comment (adammahmood) [reply repositories/PageRepository.test.js]: I agree.

File: repositories/PageRepository.js (hunk 7 above)
Comment (adammahmood) [reply repositories/PageRepository.js]: Great use of the word "nomenclature".

**Comment by samiwel** [comment]
As discussed duplicated options appear out of order. Other than that looks good and I've been testing with my front end changes seems pretty solid 👍

**Comment by tgandrews** [comment]
@samiwel this should be resolved now.

File: repositories/OptionRepository.js [hunk 11]
```diff
@@ -27,7 +27,7 @@ const checkForExistingExclusive = async answerId => {
   }
 };
 
-const findAll = (where = {}, orderBy = "created_at", direction = "asc") =>
+const findAll = (where = {}, orderBy = "id", direction = "asc") =>
```
Comment (adammahmood) [inline repositories/OptionRepository.js]: Is ordering by `id` safe? Why doesn't order by `created_at` at work?  If the `created_at` becomes the same value when duplicating. Then maybe we should order by `created_at` and then fall back onto `id`?

File: repositories/strategies/duplicateStrategy.js [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    const order = await getOrUpdateOrderForPageInsert(
+      trx,
+      returnedData.SectionId,
+      returnedData.id,
+      position
+    );
+
+    await trx
+      .table(tableName)
+      .update({ order: order })
```
Comment (adammahmood) [inline repositories/strategies/duplicateStrategy.js]: This could be `.update({order})`

Comment (adammahmood) [reply repositories/strategies/duplicateStrategy.js]: I've seen (and done this myself) a few times.  Maybe we can set up a linting rule to enforce this?

File: scripts/test.sh [hunk 13]
```diff
@@ -25,4 +25,4 @@ echo "waiting on postgres to start..."
 
 echo "running tests..."
 
-NODE_ENV=test DB_CONNECTION_URI="postgres://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/postgres" yarn jest --runInBand $1
+NODE_ENV=test DB_CONNECTION_URI="postgres://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/postgres" yarn jest --runInBand "$@"
```
Comment (adammahmood) [inline scripts/test.sh]: This should _really_ be a separate commit.

File: repositories/OptionRepository.js (hunk 11 above)
Comment (tgandrews) [reply repositories/OptionRepository.js]: When we insert in during a duplication the `created_at` date is all the same so the order could not be determined. This resolved that issue. I don't see a point in having a fall back if `id` works just as well as `created_at`.

File: repositories/strategies/duplicateStrategy.js (hunk 12 above)
Comment (tgandrews) [reply repositories/strategies/duplicateStrategy.js]: Yeah we can add it - I have definitely seen it before. I think airbnb eslint does it by default. But its a tech debt.

File: scripts/test.sh (hunk 13 above)
Comment (tgandrews) [reply scripts/test.sh]: Yes apologies. This was done debugging and I forgot to split it out. It just passes all args passed to the script down to `jest` not just the first one.

File: repositories/OptionRepository.js (hunk 11 above)
Comment (samiwel) [reply repositories/OptionRepository.js]: Ordering by `id` should be safe enough until we have a requirement to re-order options.

File: scripts/test.sh (hunk 13 above)
Comment (adammahmood) [reply scripts/test.sh]: Can we pull this out into a seperate commit?

**Comment by adammahmood** [comment]
LGTM.  Squash commits and ensure the enhancement to `scripts/test.sh` is a separate commit to the feature commit.
